### PR TITLE
Pinch Zoom

### PR DESCRIPTION
**Pull request by davidjiagoogle**
Added setZoomRatio API in CameraUseCase and PreviewViewModel, as well as its implementation in CameraXCameraUseCase.

Added pinching gesture capture in PreviewScreen

### CODE REVIEW

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt [hunk 1]
```diff
@@ -147,6 +150,21 @@ class CameraXCameraUseCase @Inject constructor(
...
+    override fun setZoomRatio(scale: Float) {
+        val zoomState = getZoomState()
+        if (zoomState != null) {
+            val finalScale = (zoomState.zoomRatio * scale).coerceIn(zoomState.minZoomRatio, zoomState.maxZoomRatio)
+            camera?.cameraControl?.setZoomRatio(finalScale)
+        }
+    }
+
+    private fun getZoomState(): ZoomState? {
+        if (camera == null) {
```
Comment (yasith) [inline domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt:162]: nit: can probably replace with `return camera?.cameraInfo.zoomState.value`

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraUseCase.kt [hunk 2]
```diff
@@ -47,4 +46,6 @@ interface CameraUseCase {
     suspend fun startVideoRecording()
 
     fun stopVideoRecording()
+
+    fun setZoomRatio(scale: Float)
```
Comment (yasith) [inline domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraUseCase.kt:50]: nit: should we call this setZoomScale ?

File: feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/PreviewScreen.kt [hunk 3]
```diff
@@ -89,7 +94,17 @@ fun PreviewScreen(
         Text(text = stringResource(R.string.camera_not_ready))
     } else if (previewUiState.cameraState == CameraState.READY) {
         Box(
-            modifier = Modifier.fillMaxSize()
+            modifier = Modifier.fillMaxSize().pointerInput(Unit) {
```
Comment (yasith) [inline feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/PreviewScreen.kt:97]: You can use the rememberTransformableState for pinch zoom gestures.

```
val transformableState = rememberTransformableState(onTransformation = { zoomChange, _, _ ->
        viewModel.setZoomRatio(zoomChange)
    })
    
...
    
modifier = Modifier.fillMaxSize().transformable(state = transformableState),
```

https://developer.android.com/jetpack/compose/touch-input/pointer-input/multi-touch

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraUseCase.kt (hunk 2 above)
Comment (davidjiagoogle) [reply domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraUseCase.kt:50]: done

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt (hunk 1 above)
Comment (davidjiagoogle) [reply domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt:162]: done

File: feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/PreviewScreen.kt (hunk 3 above)
Comment (davidjiagoogle) [reply feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/PreviewScreen.kt:97]: Done. Works like a charm

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt [hunk 4]
```diff
@@ -147,6 +150,18 @@ class CameraXCameraUseCase @Inject constructor(
         recording?.stop()
     }
 
+    override fun setZoomScale(scale: Float) {
+        val zoomState = getZoomState()
```
Comment (jsaund) [inline domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt:154]: nit: you can simplify this a bit
```
val zoomState = getZoomState() ?: return
val finalScale = ...
```

File: domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt [hunk 5]
```diff
@@ -147,6 +150,18 @@ class CameraXCameraUseCase @Inject constructor(
...
+        if (zoomState != null) {
+            val finalScale = (zoomState.zoomRatio * scale).coerceIn(zoomState.minZoomRatio, zoomState.maxZoomRatio)
+            camera?.cameraControl?.setZoomRatio(finalScale)
+        }
+    }
+
+    private fun getZoomState(): ZoomState? {
```
Comment (jsaund) [inline domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/CameraXCameraUseCase.kt:161]: nit: you can use an expression function here:
```
private fun getZoomState(): ZoomState? = camera?.cameraInfo?.zoomState?.value
```

**Review by jsaund — APPROVED** [review]
Overall LGTM!

**Pull request merged**